### PR TITLE
Allow updating SPL2 modules

### PR DESCRIPTION
**Pull request by fantavlik**
Provide functionality to input the module name and namespace for SPL2 notebooks and use that information when dispatching SPL2 (detect app and namespace from `apps.<app>` and subnamespace from `apps.<app>.my.sub.namespace`).
image
image

And provide a new button update the module contents on the connected Splunk deployment using a PUT to the `/services/spl2/modules/<namespace>.<moduleName>` endpoint:
image

Upon success an info message is provided:
image

Or an error message popup:
image

### CODE REVIEW

File: out/notebooks/commands.ts [hunk 1]
```diff
@@ -152,33 +190,72 @@ export async function enterEarliestTime(
     await vscode.workspace.applyEdit(edit);
 }
 
-export async function enterLatestTime(
-    controller: SplunkController,
-    cell: vscode.NotebookCell
-) {
+async function enterModuleName(cell: vscode.NotebookCell) {
     const cellMetadata = { ...cell.metadata };
     if (!cellMetadata.splunk) {
```
Comment (abambharoliya) [inline out/notebooks/commands.ts:195]: this check is already part of recordInput method.

Comment (abambharoliya) [reply out/notebooks/commands.ts:195]: maybe use optional chaining https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: out/notebooks/provider.ts [hunk 2]
```diff
@@ -93,6 +129,9 @@ export class CellResultCountStatusBarProvider implements vscode.NotebookCellStat
                     });
                     // earliest and latest time pickers are only supported for SPL2 at the moment
```
Comment (abambharoliya) [inline out/notebooks/provider.ts:130]: why are we having same code block at two places?

File: out/notebooks/splunk.ts [hunk 3]
```diff
@@ -56,7 +56,65 @@ export function createSearchJob(jobs, query, options) {
...
+ * @returns Promise<void> (or throw Error containing data.messages[])
+ */
+export function updateSpl2Module(service: any, moduleName: string, namespace: string, module: string) {
+    // The Splunk SDK for Javascript doesn't currently support the spl2/modules endpoints
+    // nor does it support sending requests in JSON format (only receiving responses), so
+    // for now use the underlying needle library that the SDK uses for requests/responses
+    return needle(
+        'PUT',
+        // example: https://myhost.splunkcloud.com:8089/services/spl2/modules/apps.search._default
+        `${service.prefix}/services/spl2/modules/${encodeURIComponent(namespace)}.${encodeURIComponent(moduleName)}`,
```
Comment (abambharoliya) [inline out/notebooks/splunk.ts:74]: This is weird and Redundancy! Request parameters are also part of the payload.

File: out/notebooks/splunk.ts [hunk 4]
```diff
@@ -56,7 +56,65 @@ export function createSearchJob(jobs, query, options) {
...
+        {
+            'name': moduleName,
+            'namespace': namespace,
+            'definition': module,
+        },
+        {
+            'headers': {
```
Comment (abambharoliya) [inline out/notebooks/splunk.ts:81]: The headers are mostly the same across the two calls. maybe extract it out?

File: out/notebooks/provider.ts (hunk 2 above)
Comment (fantavlik) [reply out/notebooks/provider.ts:130]: So there are two states - one for before the search was run and job was dispatched and one after, the icons are slightly different between the two. See this comment from a previous review: https://github.com/splunk/vscode-extension-splunk/pull/89/files#r1248078197

File: out/notebooks/splunk.ts (hunk 3 above)
Comment (fantavlik) [reply out/notebooks/splunk.ts:74]: I agree 😬 - I think I have some API feedback

File: out/notebooks/splunk.ts [hunk 5]
```diff
@@ -56,7 +56,65 @@ export function createSearchJob(jobs, query, options) {
...
+                'Content-Type': 'application/json',
+            },
+            'followAllRedirects': true,
+            'timeout': 0,
+            'strictSSL': false,
+            'rejectUnauthorized' : false,
+        })
+        .then((response) => {
+            const data = response.body;
+            if (!Object.prototype.isPrototypeOf(data)
```
[the remark this answers is not in the document]
Comment (fantavlik) [reply out/notebooks/splunk.ts:92]: Yes that's a good thought, I'll add that check. I do also want to retain the other conditions though to handle the case where a malformed response comes back.

File: out/notebooks/commands.ts (hunk 1 above)
Comment (fantavlik) [reply out/notebooks/commands.ts:195]: yeah good call I user optional chaining everywhere else I missed it here thanks!

**Pull request merged**